src/ghci/show_modules.rs

```diff
@@ -0,0 +1,159 @@
+                    }
+                    None => {
+                        break;
+                    }
+                }
+            }
+
+            // Remove a trailing comma, if one exists.
+            // If you have commas in your file names, G-d help you!
+            let path = token.strip_suffix(',').unwrap_or(token);
+
+            if HASKELL_SOURCE_EXTENSIONS
+                .iter()
+                .any(|extension| path.ends_with(&format!(".{extension}")))
+            {
+                source = Some(path);
+            } else {
+                outputs.push(path.into());
+            }
+        }
+
+        let source = source
+            .ok_or_else(|| {
+                miette!("Didn't find a source file in `:show modules` output line: {s:?}")
+            })?
+            .into();
+
+        Ok(Self {
+            name,
+            source,
+            outputs,
+        })
+    }
+}
+
+/// A collection of source paths, retaining information about loaded modules in a `ghci`
+/// session.
+#[derive(Debug, Clone, Default)]
+pub struct ModuleSet {
+    map: HashSet<Utf8PathBuf>,
+}
+
+impl ModuleSet {
+    /// Parse a `ModuleSet` from a set of lines read from `ghci` stdout.
+    pub fn from_lines(lines: &Lines) -> miette::Result<Self> {
+        Ok(Self {
+            map: lines
+                .iter()
+                .map(|line| {
+                    line.parse::<Module>().and_then(|module| {
+                        match module.source.canonicalize_utf8() {
+                            Ok(absolute_path) => Ok(absolute_path),
+                            Err(err) => Err(err).into_diagnostic(),
+                        }
+                    })
+                })
+                .collect::<Result<_, _>>()?,
+        })
+    }
+
+    /// Get the number of modules in this set.
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    /// Determine if a module with the given source path is contained in this module set.
+    pub fn contains_source_path(&self, path: &Utf8Path) -> bool {
+        self.map.contains(path)
+    }
+
+    /// Add a source path to this module set.
+    pub fn insert_source_path(&mut self, path: Utf8PathBuf) {
+        self.map.insert(path);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use pretty_assertions::assert_eq;
+
+    use super::*;
+
+    #[test]
+    fn test_parse_module() {
+        assert_eq!(
+            "A.MercuryPrelude ( src/A/MercuryPrelude.hs, /Users/wiggles/mwb4/dist-newstyle/build/aarch64-osx/ghc-9.6.1/mwb-0/l/test-dev/noopt/build/test-dev/A/MercuryPrelude.dyn_o )".parse::<Module>().unwrap(),
+            Module {
+                name: "A.MercuryPrelude".into(),
+                source: "src/A/MercuryPrelude.hs".into(),
+                outputs: vec!["/Users/wiggles/mwb4/dist-newstyle/build/aarch64-osx/ghc-9.6.1/mwb-0/l/test-dev/noopt/build/test-dev/A/MercuryPrelude.dyn_o".into()],
```

Comment:
you might might want to remove internal module names from this example if you plan to open source this

Reply:
The repo has been public this whole time >:)

I don't think it's a secret that we have a backend and a prelude module :p